crates/edr_instrument/src/coverage.rs

```diff
@@ -86,32 +86,108 @@ pub fn instrument_code(
     let mut metadata = Vec::new();
 
     let mut statement_counter: u64 = 0;
+
+    // Calculate metadata for statements using the correct text ranges
     let mut cursor = parsed_file.create_tree_cursor();
-    while cursor.go_to_next() {
-        match cursor.node() {
-            Node::Nonterminal(node) if node.kind == NonterminalKind::Statement => {
-                statement_counter += 1;
-
-                let hash = compute_deterministic_hash_for_statement(
-                    source_id,
-                    &content_hash,
-                    statement_counter,
-                );
-
-                instrumented_source.push_str(&format!("__HardhatCoverage.sendHit({hash}); "));
-
-                let range = cursor.text_range();
-                metadata.push(InstrumentationMetadata {
-                    tag: hash,
-                    kind: "statement",
-                    start_utf16: range.start.utf16,
-                    end_utf16: range.end.utf16,
-                });
-            }
-            Node::Terminal(node) => {
-                instrumented_source.push_str(&node.text);
+    while !cursor.is_completed() {
+        if let NodeKind::Nonterminal(NonterminalKind::Statement) = cursor.node().kind() {
+            statement_counter += 1;
+
+            let hash = compute_deterministic_hash_for_statement(
+                source_id,
+                &content_hash,
+                statement_counter,
+            );
+
+            let text_range = cursor.text_range();
+            metadata.push(InstrumentationMetadata {
+                tag: hash,
+                kind: "statement",
+                start_utf16: text_range.start.utf16,
+                end_utf16: text_range.end.utf16,
+            });
+        }
+
+        cursor.go_to_next();
+    }
+
+    let mut metadata_iter = metadata.iter();
+    let mut queue = vec![parsed_file.create_tree_cursor()];
+    while let Some(mut cursor) = queue.pop() {
+        while !cursor.is_completed() {
+            match cursor.node() {
+                Node::Nonterminal(node) => {
+                    if node.kind == NonterminalKind::Statement {
+                        let parent = cursor
+                            .ancestors()
+                            .next()
+                            .expect("Cursor should have a parent");
+
+                        if matches!(
+                            parent.kind,
+                            NonterminalKind::IfStatement
+                                | NonterminalKind::ForStatement
+                                | NonterminalKind::WhileStatement
+                                | NonterminalKind::DoWhileStatement
+                                | NonterminalKind::ElseBranch
```

Comment:
I'd move this one next to `IfStatement`.

Reply:
Addressed by https://github.com/NomicFoundation/edr/pull/918/commits/4a776ae90e6a0d02d815301fa0d5ed8687896d91